### ISSUE TITLE
Type mismatch with static[Slice[int]]

### ISSUE THREAD

**Issue by andreaferretti** [nim-lang/Nim#8524]
Let us define a type to hold a sequence of ints and also track the size at compile time:

```nim
type
  IntArray[N: static[int]] = array[N, int]
  Vector[N: static[int]] = distinct seq[int]

proc vector[N: static[int]](a: IntArray[N]): Vector[N] = Vector[N](@a)

let x = vector([1, 2, 3, 4, 5])
```

So far, so good. Now we try to define a slicing operation:

```nim
proc `[]`[N: static[int]](v: Vector[N], s: static[Slice[int]]): Vector[len(s)] =
  Vector[len(s)](seq[int](v)[s])

let y = x[2 .. 3]
```

When we do this, we get the error

```
example.nim(12, 10) Error: type mismatch: got <Vector[5], static[HSlice[system.int, system.int]]((a: 2, b: 3))>
but expected one of: 
proc `[]`[N: static[int]](v: Vector[N]; s: static[Slice[int]]): Vector[len(s)]
proc `[]`[I: Ordinal; T](a: T; i: I): T
proc `[]`[T, U](s: string; x: HSlice[T, U]): string
proc `[]`[Idx, T](a: var array[Idx, T]; i: BackwardsIndex): var T
proc `[]`[Idx, T, U, V](a: array[Idx, T]; x: HSlice[U, V]): seq[T]
proc `[]`[T](s: openArray[T]; i: BackwardsIndex): T
proc `[]`[T](s: var openArray[T]; i: BackwardsIndex): var T
proc `[]`[Idx, T](a: array[Idx, T]; i: BackwardsIndex): T
proc `[]`(s: string; i: BackwardsIndex): char
proc `[]`[T, U, V](s: openArray[T]; x: HSlice[U, V]): seq[T]

expression: [](x, 2 .. 3)
```

Notice that the error message says that we got `<Vector[5], static[HSlice[system.int, system.int]]((a: 2, b: 3))>` and there is a candidate implementation `proc [][N: static[int]](v: Vector[N]; s: static[Slice[int]]): Vector[len(s)]` that should match but doesn't.

Even changing the signature of `[]` to the more explicit

```nim
proc `[]`[N: static[int]](v: Vector[N], s: static[HSlice[int, int]]): Vector[len(s)]
```

does not improve this

**Comment by exelotl** [comment]
I just ran into this one myself. Very minimal example:

```nim
macro foo(s:static[HSlice[int,int]]) =
  echo s

foo(1..5)
```

Output (using Nim 1.2.4 on the playground)

```
/usercode/in.nim(6, 4) Error: type mismatch: got <static[HSlice[system.int, system.int]]((a: 1, b: 5))>
but expected one of: 
macro foo(s: static[HSlice[int, int]])
  first type mismatch at position: 1
  required type for s: static[HSlice[system.int, system.int]]
  but expression '1 .. 5' is of type: static[HSlice[system.int, system.int]]((a: 1, b: 5))

expression: foo(1 .. 5)
```

Note, I also get the same error if I use the more explicit syntax
```nim
foo(HSlice[int,int](a:1, b:5))
```
And also if I try using a constant:
```nim
const h: HSlice[int,int] = 1..5
foo(h)
```

### Workaround

In my case, I was able to work around it by using a wrapper template:

```nim
macro fooImpl(a, b: static[int]) =
  echo a..b

template foo(s:Slice[int]) =
  const ss = s  # Ensure that `s` is a valid compile-time expression
  fooImpl(ss.a, ss.b)

foo(1..5)
```

**Comment by cooldome** [comment]
Fixed already, covered by tests already

**Comment by exelotl** [comment]
Nice, I tested in the 1.4.2 RC and it works fine now. :slightly_smiling_face: 
For reference, this issue was a specific case of #11142 and was fixed by #15847
